std/assembly/array.ts

```diff
@@ -493,6 +493,68 @@ export class Array<T> {
     return <string>unreachable();
   }
 
+  flat(): T {
+    if (!isArray<T>()) {
+      ERROR("Cannot call flat() on Array<T> where T is not an Array.");
```

Comment:
I'm not sure we should restrict this. What about just return shallow clone of array? Btw it will be follow spec

Reply:
This is impossible because the return type must be known at compile time.

`valueof<T>[]` will throw an error, and we cannot be spec compliant in this way.

Reply:
```ts
if (!isArray<T>) {
   return this.slice();
}
```
?

Reply:
In the case of `u8[]`, calling flat will return `u8`, and not `u8[]`.

Returning `this.slice() as T` will return `u8` instead of `u8[]`.

Reply:
@dcodeIO any thoughts?

Reply:
Ideally it should define as:
```ts
flat<U>(this: U[]): U[] {
   ...
}
```
But I'm not sure it support for now

Reply:
we need to support conditional return types for this to work. If `T` isn't an array, then it should return `Array<T>` instead of `T`.

Reply:
Suggesting to leave as-is until more work is done on the compiler directly.